README.md
# vznncv-geekwork-x735-v2.5

Debian package version of [geekwork-x735-v2.5](https://github.com/geekworm-com/x735-v2.5) scripts.

![geekwork-x735-v2.5](docs/geekworm-x735-v2.5.jpg)

Unlike original scripts it has the following differences:

- safe shutdown script is integrated as special *systemd* service and is activated automatically by default `shutdown`
  command.
- FAN uses hardware PWM implementation.
- FAN start/stop parameters can be adjusted via '/etc/geekworm/config.conf' configuration.

## Installation

1. Download package from [github releases](https://github.com/vznncv/vznncv-geekworm-x735-v2.5/releases) section
   or build package locally with `package/build_debian_package.sh` script.

2. Copy and install package to raspberry pi:

   ```shell
   sudo dpkg -i geekworm-x735-v2.5_<package_version>_all.deb
   sudo apt-get install --fix-broken
   ```

## Uninstallation

```
sudo apt-get remove geekworm-x735-v2.5
```

### Power consumption optimization

Fan script uses [pigpiod](https://abyz.me.uk/rpi/pigpio/pigpiod.html), that consumes 4-6% of CPU in idle state.
To reduce CPU usage you can disable alert functionality of *pigpiod*, as it isn't requires by fan scripts:

1. Edit file `/usr/lib/systemd/system/pigpiod.service` to add `-m -s 10` to daemon startup options:

   ```
   # ...
   [Service]
   ExecStart=/usr/bin/pigpiod -l -m -s 10
   # ...
   ```

2. Reload systemd and restart service:

   ```
   sudo systemctl daemon-reload
   sudo systemctl restart pigpiod
   ```

### Configuration

If you need to check fan settings (minimal temperature to start fan, etc.), you need to edit
`/etc/geekworm/config.conf` configuration file and restart fan service:

```shell
sudo systemctl restart geekworm-x735-fan
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.2.0] - 2021-12-16
### Added
- Added `/etc/geekworm-x735/config.conf` file with fan configuration options.
### Changed
- Refactor `geekworm-x735-power` script to use *sysfs* instead of *pigpio*. It helps to avoid problems with
  network availability during shutdown.
- Switch GPIO input pull up/down mode to output pull/push for power control pins to simplify `geekworm-x735-power`
  script implementation.
- Refactor debian package build scripts to use `dpgk-buildpackage` instead of `dpkg-deb`
### Fixed
- Fixed systemd safe shutdown/reboot hooks.


## [0.1.0] - 2021-09-12
### Added
- Added shield fan control service
- Added shield power signal monitor service
- Added safe shutdown service


geekworm-x735-fan.py
#!/usr/bin/env python3
"""
Helper script to control geekwork x735 PWM FAN
"""
import argparse
import configparser
import logging
import os.path
import time
from collections import namedtuple
from typing import Optional

import pigpio

logger = logging.getLogger("fan-service")


def _build_number_parser(min=None, max=None, optional=False):
    def validator(value):
        if value is None or value == '':
            if optional:
                return None
            else:
                raise ValueError(f"value mustn't be empty, but it was: {value}")

        try:
            parsed_value = float(value)
        except ValueError:
            raise ValueError(f"value should number, but it was {value}")

        if min is not None and parsed_value < min:
            raise ValueError(f"value should greater or equal than {min}, but it was {value}")
        if max is not None and parsed_value > max:
            raise ValueError(f"value should less or equal than {max}, but it was {value}")

        return parsed_value

    return validator


_OptionInfo = namedtuple('_OptionInfo', ('name', 'description', 'default_value', 'type'))


class _Config:
    _CONFIG_DESCRIPTION = {
        'pigpio': [
            _OptionInfo('port', 'pigpio Daemon port', None, _build_number_parser(min=0, optional=True)),
        ],
        'fan': [
            _OptionInfo('update_period', 'FAN update period', 2, _build_number_parser(min=0)),
            _OptionInfo('log_period', 'FAN state log peroid', 60, _build_number_parser(min=0)),

            _OptionInfo('duty_cycle_min', 'minimal pwm duty cycle', 0.2, _build_number_parser(min=0, max=1)),
            _OptionInfo('duty_cycle_max', 'maximal pwm duty cycle', 1.0, _build_number_parser(min=0, max=1)),
            _OptionInfo('min_power_start_temp', 'minimal temperature to start FAN', 48, _build_number_parser()),
            _OptionInfo('min_power_stop_temp', 'minimal temperature to stop FAN', 42, _build_number_parser()),
            _OptionInfo('max_power_temp', 'temperature that corresponds max FAN power', 60, _build_number_parser()),
        ]
    }

    @classmethod
    def add_cli_args(cls, parser: argparse.ArgumentParser):
        for section_name, section_options in cls._CONFIG_DESCRIPTION.items():
            for option in section_options:
                name = f'--{section_name}-{option.name}'.lower().replace('_', '-')
                dest = f'{section_name}_{option.name}'
                parser.add_argument(name, default=None, dest=dest, help=option.description)
        return parser

    _DEFAULT_CONFIG_PATH = '/etc/geekworm-x735/config.conf'

    def __init__(self, path, parser_options: Optional[argparse.Namespace] = None):

        # load options from configuration file
        config = configparser.ConfigParser()
        if path:
            config.read(path)
        elif os.path.exists(self._DEFAULT_CONFIG_PATH):
            config.read(self._DEFAULT_CONFIG_PATH)
        else:
            logger.warning(f"No config file is specify and default config file {self._DEFAULT_CONFIG_PATH}"
                           f"isn't found. Default values for all options will be used.")
        raw_config = {}
        for section_name, section_config in config.items():
            if section_name == configparser.DEFAULTSECT:
                continue
            if section_name not in self._CONFIG_DESCRIPTION:
                logger.error(f"Unknown section [{section_name}]")
                continue
            section_options = self._CONFIG_DESCRIPTION[section_name]

            raw_section_config = raw_config[section_name] = {}
            section_config_dict = {k: v for k, v in section_config.items()}
            for option in section_options:
                if option in section_config_dict:
                    raw_section_config[option.name] = section_config_dict.pop(option.name)
            if section_config_dict:
                raise ValueError(f"Unknown options of the section [{section_name}]: {section_config_dict}")
        # add options from parser
        for section_name, section_options in self._CONFIG_DESCRIPTION.items():
            for option in section_options:
                dest = f'{section_name}_{option.name}'
                value = getattr(parser_options, dest, None)
                if value is not None:
                    raw_config.setdefault(section_name, {})[option.name] = value

        # parse/validate options
        self._parsed_config = {}
        for section_name, section_options in self._CONFIG_DESCRIPTION.items():
            parsed_section_config = self._parsed_config[section_name] = {}
            raw_section_config = raw_config.get(section_name, {})
            for option in section_options:
                raw_value = raw_section_config.get(option.name, option.default_value)
                try:
                    parsed_section_config[option.name] = option.type(raw_value)
                except Exception as e:
                    raise ValueError(f"Fail to process option {section_name}/{option.name}") from e
        if not self.min_power_stop_temp <= self.min_power_start_temp <= self.max_power_temp:
            raise ValueError(
                f"Temperature should be min_power_stop_temp <= min_power_start_temp <= max_power_temp, "
                f"but it was: {self.min_power_stop_temp} - {self.min_power_start_temp} - {self.max_power_temp}"
            )
        if not self.duty_cycle_min <= self.duty_cycle_max:
            raise ValueError(
                f"Duty cycle should be duty_cycle_min <= duty_cycle_max, "
                f"but it was: {self.duty_cycle_min} - {self.duty_cycle_max}"
            )

    @property
    def update_period(self) -> float:
        return self._parsed_config['fan']['update_period']

    @property
    def log_period(self) -> float:
        return self._parsed_config['fan']['log_period']

    @property
    def duty_cycle_min(self) -> float:
        return self._parsed_config['fan']['duty_cycle_min']

    @property
    def duty_cycle_max(self) -> float:
        return self._parsed_config['fan']['duty_cycle_max']

    @property
    def min_power_start_temp(self) -> float:
        return self._parsed_config['fan']['min_power_start_temp']

    @property
    def min_power_stop_temp(self) -> float:
        return self._parsed_config['fan']['min_power_stop_temp']

    @property
    def max_power_temp(self) -> float:
        return self._parsed_config['fan']['max_power_temp']

    @property
    def pigpio_port(self) -> Optional[int]:
        return self._parsed_config['pigpio']['port']

    def log_params(self):
        logger.info("Effective service options:")
        for section_name, section_config in self._parsed_config.items():
            for option_name, value in section_config.items():
                logger.info(f"- {section_name}/{option_name}: {value}")


class _TemperatureManager:
    _TEMP_SYS_PATH = '/sys/class/thermal/thermal_zone0/temp'

    def get_cpu_temp(self) -> float:
        with open(self._TEMP_SYS_PATH, 'rb') as f:
            raw_temp = f.read().strip()
        return int(raw_temp) / 1000.0


class _X735FANManager:
    FAN_PWM_GPIO = 13
    FAN_PWM_FREQ = 25000

    def __init__(self, config):
        self._temp_manager = _TemperatureManager()
        self._config = config
        pi_args = {}
        if self._config.pigpio_port:
            pi_args['port'] = self._config.pigpio_port
        self._pi = pigpio.pi(**pi_args)

        # previous duty cycle
        self._prev_duty_cycle = 0.0

    def close(self):
        self._pi.stop()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()

    def _update_duty_cycle(self, temp: float) -> float:
        if temp <= self._config.min_power_stop_temp:
            result = 0
        elif temp <= self._config.min_power_start_temp:
            if self._prev_duty_cycle > 0:
                result = self._config.duty_cycle_min
            else:
                result = 0
        elif temp <= self._config.max_power_temp:
            k = (temp - self._config.min_power_start_temp) / \
                (self._config.max_power_temp - self._config.min_power_start_temp)
            result = self._config.duty_cycle_min + (self._config.duty_cycle_max - self._config.duty_cycle_min) * k
        else:
            result = self._config.duty_cycle_max
        self._prev_duty_cycle = result
        return result

    @staticmethod
    def _to_pigpio_hw_dutycycle(value):
        value = int(value * 1_000_000)
        if value < 0:
            value = 0
        elif value > 1_000_000:
            value = 1_000_000
        return value

    def _fan_enable(self):
        self._pi.set_mode(self.FAN_PWM_GPIO, pigpio.OUTPUT)

    def _fan_disable(self):
        self._pi.set_mode(self.FAN_PWM_FREQ, pigpio.INPUT)

    def _fan_set_duty_cycle(self, value):
        self._pi.hardware_PWM(self.FAN_PWM_GPIO, self.FAN_PWM_FREQ, self._to_pigpio_hw_dutycycle(value))

    def main_loop(self):
        logger.info(f"Start x735 temperature/fan monitor")
        next_log = time.time()

        cpu_temp = self._temp_manager.get_cpu_temp()
        duty_cycle = 0

        try:
            # configure fan pin
            self._fan_enable()

            while True:
                cpu_temp = self._temp_manager.get_cpu_temp()
                duty_cycle = self._update_duty_cycle(cpu_temp)
                self._fan_set_duty_cycle(duty_cycle)
                if next_log <= time.time():
                    next_log += self._config.log_period
                    logger.info(f"CPU temperature: {cpu_temp:.2f}. Fan: {duty_cycle:.1%}")

                time.sleep(self._config.update_period)
        except KeyboardInterrupt:
            # force log of last cpu temperature/duty cycle
            logger.info(f"CPU temperature: {cpu_temp:.2f}. Fan: {duty_cycle:.1%}")
            raise
        finally:
            # disable pwm
            self._fan_set_duty_cycle(0)
            # disable pin
            self._fan_disable()


def _main_loop(config):
    while True:
        try:
            with _X735FANManager(config=config) as manager:
                manager.main_loop()
        except Exception:
            logger.exception("Fan service error. Retry in 30 seconds ...")
        time.sleep(30)


def main(args=None):
    parser = argparse.ArgumentParser(description='geekwork x735 fan manager')
    parser.add_argument('-c', '--config', help='Config file')
    _Config.add_cli_args(parser)
    parsed_args = parser.parse_args(args)
    logging.basicConfig(level=logging.INFO, format='%(asctime)s  [%(name)s:%(lineno)s]  %(levelname)s - %(message)s')

    config = _Config(parsed_args.config, parser_options=parsed_args)
    config.log_params()
    try:
        _main_loop(config)
    except KeyboardInterrupt:
        logger.info("Shutdown ...")


if __name__ == '__main__':
    main()


package/build_debian_package.sh
#!/usr/bin/env bash
set -e
SCRIPT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
SCRIPT_NAME="$(basename "$0")"
SOURCE_DIR="$(dirname "${SCRIPT_DIR}")"

#
# Helper functions
#

function log() {
    echo "$*" >&2
}
function log_info() {
    log "INFO: $*"
}

function log_error() {
    log "ERROR: $*"
}

function get_debian_control_field() {
    local field_name="$1"
    local field_value
    field_value=$(sed -n -E "s/^${field_name}:\s*([^\s]+)\s*$/\1/p" <"$SCRIPT_DIR/debian/control")
    if [[ -z "$field_value" ]]; then
        log_error "Cannot extract package \"${field_name}\" from debian/control"
        exit 1
    fi
    echo "$field_value"
}

function assert_tool() {
    local tool_name="$1"
    if ! which "$tool_name" >/dev/null; then
        log_error "$tool_name is required, but isn't found!"
        exit 1
    fi
}

#
# CLI interface
#

function help() {
    log "Helper script to build debian package"
    log "Usage: $SCRIPT_NAME [-h|--help] [--dev-build)]"
}
DEV_BUILD=""
while [[ $# -gt 0 ]]; do
    case "$1" in
    -h | --help)
        help
        exit 0
        ;;
    --dev-build)
        DEV_BUILD="1"
        shift
        ;;
    *)
        log "ERROR: unknown option: $1"
        exit 1
        ;;
    esac
done

#
# main logic
#

cd "$SCRIPT_DIR"

# 1. build debian changelog from CHANGELOG.md
log_info "Generate dummy debian/changelog"
# read current maintainer
PACKAGE_MAINTAINER="$(sed -n -E -e 's/Maintainer:\s+(.*)/\1/p' debian/control)"
# read package name
PACKAGE_NAME="$(sed -n -E -e 's/Package:\s+(.*)/\1/p' debian/control)"
SOURCE_NAME="$(sed -n -E -e 's/Source:\s+(.*)/\1/p' debian/control)"
# read versions and release dates from CHANGELOG.md
readarray -t RELEASED_VERSIONS <<<"$(sed -n -E -e 's/^##\s*\[([0-9.]*)\][^0-9]*([0-9-]+)\s*$/\1:\2/p' "${SOURCE_DIR}/CHANGELOG.md")"
# add dummy release version if no versions are released
if [[ "${#RELEASED_VERSIONS[@]}" -eq 0 ]]; then
    RELEASED_VERSIONS+=("0.1.0-1:$(date '+%Y-%m-%d')")
fi
# override version (it can be used for dev package building)
if [[ "$DEV_BUILD" == 1 ]]; then
    LAST_VERSION="${RELEASED_VERSIONS[0]%:*}"
    NEXT_VERSION="$((${LAST_VERSION%%.*} + 1)).0.0-dev"
    RELEASED_VERSIONS=("$NEXT_VERSION:$(date '+%Y-%m-%d')" "${RELEASED_VERSIONS[@]}")
fi
# build changelog
DEBIAN_CHANGELOG_FILE="debian/changelog"
for RELEASE_IFNO in "${RELEASED_VERSIONS[@]}"; do
    RELEASE_VERSION="${RELEASE_IFNO%:*}"
    RELEASE_DATE="${RELEASE_IFNO#*:}"
    echo "${SOURCE_NAME} (${RELEASE_VERSION}) stable; urgency=medium"
    echo ""
    echo "  * Upstream release"
    echo ""
    echo "-- ${PACKAGE_MAINTAINER} $(date -d "$RELEASE_DATE" -u -R)"
    echo ""
done | head -n -1 >"${DEBIAN_CHANGELOG_FILE}"

# 2. run dpkg-buildpackage
log_info "Build package"
assert_tool dpkg-buildpackage
dpkg-buildpackage --build=binary --unsigned-source --unsigned-changes

## 3. move artifacts to dist
log_info "Move artifacts to dist directory"
OUTPUT_DIR="${SCRIPT_DIR}/dist"
if [[ -e "$OUTPUT_DIR" ]]; then
    rm -rf "$OUTPUT_DIR"
fi
mkdir "$OUTPUT_DIR"
mv "../${PACKAGE_NAME}_"* "$OUTPUT_DIR"
mv "../${SOURCE_NAME}_"* "$OUTPUT_DIR"

log_info "Complete!"


geekworm-x735-v2.5-fan-service.py
#!/usr/bin/env python3
import argparse
import logging
import time

import pigpio

_FAN_PWM_UPDATE_PERIOD = 2
_LOG_UPDATER_PERIOD = 5 * 60
logger = logging.getLogger("fan-service")

_FAN_PWM_PIN = 13
_TEMP_SYS_PATH = '/sys/class/thermal/thermal_zone0/temp'


def _get_cpu_temp():
    with open(_TEMP_SYS_PATH, 'rb') as f:
        raw_temp = f.read().strip()
    return int(raw_temp) / 1000.0


_CPU_FAN_TEMP_MIN = 40
_CPU_FAN_TEMP_MIN_DUTY_CYCLE = 0.2
_CPU_FAN_TEMP_MAX = 65
_CPU_FAN_PWM_FREQ = 25000


class _FANController:
    _CPU_FAN_TEMP_MIN_START = 55
    _CPU_FAN_TEMP_MIN_STOP = 45

    _CPU_FAN_DS_MIN = 0.2
    _CPU_FAN_PWM_FREQ = 25000

    _CPU_FAN_TEMP_MIN = 45
    _CPU_FAN_TEMP_MAX = 65

    def __init__(self):
        self._activate = False

    def get_duty_cycle(self, temp):
        duty_cycle = (temp - self._CPU_FAN_TEMP_MIN) / (self._CPU_FAN_TEMP_MAX - self._CPU_FAN_TEMP_MIN)
        if duty_cycle > 1.0:
            duty_cycle = 1.0
        elif duty_cycle < self._CPU_FAN_DS_MIN:
            duty_cycle = self._CPU_FAN_DS_MIN

        if temp > self._CPU_FAN_TEMP_MIN_START:
            self._activate = True
        elif temp > self._CPU_FAN_TEMP_MIN_STOP:
            if not self._activate:
                duty_cycle = 0.0
        else:
            self._activate = False
            duty_cycle = 0.0

        return duty_cycle


def _cpu_temp_to_duty_cycle(temp):
    if temp <= _CPU_FAN_TEMP_MIN:
        return 0.0
    else:
        duty_cycle = (temp - _CPU_FAN_TEMP_MIN) / (_CPU_FAN_TEMP_MAX - _CPU_FAN_TEMP_MIN)
        duty_cycle = duty_cycle * (1 - _CPU_FAN_TEMP_MIN_DUTY_CYCLE) + _CPU_FAN_TEMP_MIN_DUTY_CYCLE

    if duty_cycle < 0:
        duty_cycle = 0.0
    elif duty_cycle > 1:
        duty_cycle = 1.0
    return duty_cycle


def _to_pigpio_hw_dutycycle(value):
    value = int(value * 1_000_000)
    if value < 0:
        value = 0
    elif value > 1_000_000:
        value = 1_000_000
    return value


def main(args=None):
    parser = argparse.ArgumentParser(description='geekwork x735 hat fan daemon')
    parsed_args = parser.parse_args(args)
    logging.basicConfig(level=logging.INFO, format='%(asctime)s  [%(name)s:%(lineno)s]  %(levelname)s - %(message)s')

    pi = pigpio.pi()
    # configure fan pin
    pi.set_mode(_FAN_PWM_PIN, pigpio.OUTPUT)

    logger.info(f"Start ...")
    next_log = time.time()

    fan_controller = _FANController()

    try:
        while True:
            cpu_temp = _get_cpu_temp()
            duty_cycle = fan_controller.get_duty_cycle(cpu_temp)
            pi.hardware_PWM(_FAN_PWM_PIN, _CPU_FAN_PWM_FREQ, _to_pigpio_hw_dutycycle(duty_cycle))
            if time.time() >= next_log:
                logger.info(f"CPU: {cpu_temp:.1f} C; Fan: {duty_cycle:.1%}")
                next_log += _LOG_UPDATER_PERIOD

            time.sleep(_FAN_PWM_UPDATE_PERIOD)
    except KeyboardInterrupt:
        logger.info("Shutdown ...")
    finally:
        cpu_temp = _get_cpu_temp()
        logger.info(f"Stop. CPU: {cpu_temp:.1f} C; Fan: ${0:.1%} (disable)")
        # disable pwm
        pi.hardware_PWM(_FAN_PWM_PIN, _CPU_FAN_PWM_FREQ, 0)
        # disable pin
        pi.set_mode(_FAN_PWM_PIN, pigpio.INPUT)


if __name__ == '__main__':
    main()


geekworm-x735-power.py
#!/usr/bin/env python3
"""
Helper script to control power shield switch on/off.
"""
import argparse
import logging
import os
import subprocess
import time

logger = logging.getLogger("power-service")


#
# Minimal native GPIO control implementation.
#

class _SYSFSGPIOManager:
    _SYSFS_GPIO_PATH = '/sys/class/gpio'

    def __init__(self):
        pass

    @classmethod
    def _check_pin(cls, pin):
        if not 0 <= pin < 54:
            raise ValueError(f"Unknown pin number {pin}")

    def _export_and_get_pin_path(self, pin):
        self._check_pin(pin)
        gpio_path = os.path.join(self._SYSFS_GPIO_PATH, f'gpio{pin}')
        if os.path.exists(gpio_path):
            return gpio_path
        # try to export pin
        with open(os.path.join(self._SYSFS_GPIO_PATH, 'export'), 'w') as f:
            f.write(str(pin))
        if not os.path.exists(gpio_path):
            raise ValueError(f"Fail to activate pin {pin}")
        return gpio_path

    def _gpio_direction_path(self, pin):
        return os.path.join(self._export_and_get_pin_path(pin), 'direction')

    def _gpio_value_path(self, pin):
        return os.path.join(self._export_and_get_pin_path(pin), 'value')

    def get_mode(self, pin):
        gpio_direction_path = self._gpio_direction_path(pin)
        with open(gpio_direction_path) as f:
            value = f.read().strip()
        if value == "in":
            return "input"
        elif value == "out":
            return "output"
        else:
            raise ValueError(f"Unknown pin mode: {value}")

    def set_mode(self, pin, mode):
        gpio_direction_path = self._gpio_direction_path(pin)
        if mode == "input":
            value = "in"
        elif mode == "output":
            value = "out"
        else:
            raise ValueError(f"Unknown mode: {mode}")
        prev_value = self.get_mode(pin)
        if prev_value != value:
            with open(gpio_direction_path, 'w') as f:
                return f.write(value)

    def set_value(self, pin, value):
        gpio_value_path = self._gpio_value_path(pin)
        with open(gpio_value_path, 'w') as f:
            f.write(str(value))

    def get_value(self, pin):
        gpio_value_path = self._gpio_value_path(pin)
        with open(gpio_value_path, 'r') as f:
            raw_value = f.read()
        return int(raw_value.strip())


#
# X735 power manager
#

class _PowerManagerX735:
    # boot indication pin
    BOOT_INFO_GPIO = 12
    # shutdown control pin
    SHUTDOWN_CONTROL_GPIO = 20
    # shutdown signal pin
    SHUTDOWN_SIGNAL_GPIO = 5

    def __init__(self):
        self.gpio = _SYSFSGPIOManager()

    def _activate_pin(self, pin, name, mode, value=0):
        prev_mode = self.gpio.get_mode(pin)
        prev_value = self.gpio.get_value(pin)
        if mode != prev_mode or (mode != 'input' and value != prev_value):
            logger.info(f"Activate/reset {name} pin ({pin}, mode: {mode})")
            self.gpio.set_mode(pin, mode)
            if mode != 'input':
                self.gpio.set_value(pin, value)

    def init_pins(self):
        """
        Reset power control pins to default values.
        """
        # reset shutdown control pin
        self._activate_pin(self.SHUTDOWN_CONTROL_GPIO, 'showdown control', 'output', 0)
        # reset shutdown info pin
        self._activate_pin(self.SHUTDOWN_SIGNAL_GPIO, 'shutdown signal', 'input')
        # activate indication pin
        self._activate_pin(self.BOOT_INFO_GPIO, 'boot info', 'output', 1)

    def _send_shutdown_signal(self, signal_name, pulse_length, post_pulse_delay=1):
        # ensure that pins are initialized
        self.init_pins()

        # send signal
        logger.info(f"Send {signal_name} signal to x735 ...")
        self.gpio.set_value(self.SHUTDOWN_CONTROL_GPIO, 1)
        time.sleep(pulse_length)
        self.gpio.set_value(self.SHUTDOWN_CONTROL_GPIO, 0)
        time.sleep(post_pulse_delay)

        logger.info("Complete")

    def safe_poweroff(self):
        """
        Safe shutdown.
        """
        self._send_shutdown_signal(signal_name="shutdown", pulse_length=4)

    def safe_reboot(self):
        """
        Safe reboot.
        """
        self._send_shutdown_signal(signal_name="reboot", pulse_length=1)

    SHUTDOWN_SIGNAL_LOW_PULL_PERIOD = 0.2
    SHUTDOWN_SIGNAL_HIGH_PULL_PERIOD = 0.02

    SHUTDOWN_SIGNAL_REBOOT_PULSE = 0.2
    SHUTDOWN_SIGNAL_POWEROFF_PULSE = 0.6

    SHUTDOWN_TIMEOUT = 30

    def monitor(self):
        """
        Monitor shutdown/reboot signal.
        """
        try:

            while True:
                self._monitor_loop()
                time.sleep(self.SHUTDOWN_TIMEOUT)
                logger.warning(
                    f"System haven't been shut down after {self.SHUTDOWN_TIMEOUT} seconds. Continue monitoring!"
                )
        except KeyboardInterrupt:
            logger.info("Shutdown ...")

    def _shutdown_system(self, shutdown_command):
        result = subprocess.run(['systemctl', 'is-system-running'], encoding='utf-8', stdout=subprocess.PIPE)
        state = result.stdout.strip()
        # don't call any shutdown command if system is stopping
        if state == 'stopping':
            return
        if shutdown_command == 'poweroff':
            subprocess.check_output(['shutdown', '--poweroff', 'now'])
        elif shutdown_command == 'reboot':
            subprocess.check_output(['shutdown', '--reboot', 'now'])
        else:
            raise ValueError(f"Unknown shutdown command: {shutdown_command}")

    def _monitor_loop(self):
        """
        Monitor shutdown/reboot signal.
        """
        # ensure that pins are initialized
        self.init_pins()

        logger.info(f"Start power control pin ({self.SHUTDOWN_SIGNAL_GPIO}) monitoring")

        while True:
            time.sleep(self.SHUTDOWN_SIGNAL_LOW_PULL_PERIOD)
            shutdown_signal = self.gpio.get_value(self.SHUTDOWN_SIGNAL_GPIO)
            if not shutdown_signal:
                continue

            logger.info("Detect pulse start")
            pulse_start = time.time()
            time.sleep(self.SHUTDOWN_SIGNAL_HIGH_PULL_PERIOD)

            while True:
                pulse_time = time.time() - pulse_start
                pulse_time_ms = int(pulse_time * 1000)
                shutdown_signal = self.gpio.get_value(self.SHUTDOWN_SIGNAL_GPIO)
                if not shutdown_signal:
                    break

                if pulse_time > self.SHUTDOWN_SIGNAL_POWEROFF_PULSE:
                    # reboot pulse
                    logger.info(f"Shutting down (pulse time >= {pulse_time_ms} ms) ...")
                    self._shutdown_system('poweroff')
                    return
                time.sleep(self.SHUTDOWN_SIGNAL_HIGH_PULL_PERIOD)

            if pulse_time > self.SHUTDOWN_SIGNAL_REBOOT_PULSE:
                logger.info(f"X735 Rebooting (pulse time {int(pulse_time * 1000)} ms)...")
                self._shutdown_system('reboot')
                return
            else:
                logger.info(f"Invalid control pulse ({int(pulse_time * 1000)} ms)")


#
# CLI interface
#

def main(args=None):
    parser = argparse.ArgumentParser(description='geekwork x735 hat fan daemon')
    parser.add_argument('command', help="power command", choices=[
        'monitor',
        'safe-shutdown', 'safe-poweroff',
        'safe-reboot'
    ])
    parsed_args = parser.parse_args(args)
    logging.basicConfig(level=logging.INFO, format='%(asctime)s  [%(name)s:%(lineno)s]  %(levelname)s - %(message)s')

    if os.geteuid() != 0:
        raise ValueError("script should be run with root permissions")

    power_manager = _PowerManagerX735()

    if parsed_args.command == 'monitor':
        power_manager.monitor()
    elif parsed_args.command == 'safe-shutdown' or parsed_args.command == 'safe-poweroff':
        power_manager.safe_poweroff()
    elif parsed_args.command == 'safe-reboot':
        power_manager.safe_reboot()
    else:
        raise ValueError(f"Unknown command: {parsed_args.command}")


if __name__ == '__main__':
    main()
